### ISSUE TITLE
Boxed types as primitives

### ISSUE THREAD

**Issue by Thrameos** [jpype-project/jpype#536]
@marscher  I am trying to improve the speed in the JPype 0.8.  Unfortunately, I have run into an issue that would break compatible.  Currently boxed types inherit from primitives so that they will function anywhere that requires a number type without requiring a forced conversion.  This is forcing everything is to go through a type system which pulls in a huge number of python calls for each routine.  I can remove this restriction, but the isinstance will no longer work for boxed types.  I will have to make them work like numbers rather than being int or floats.

Any opinions?

**Comment by marscher** [comment]
Eventually you circumvent the type hierarchy by implementing a custom instance check. https://docs.python.org/3/reference/datamodel.html#customizing-instance-and-subclass-checks

Am 16. November 2019 23:55:51 MEZ schrieb Karl Nelson <notifications@github.com>:
>@marscher  I am trying to improve the speed in the JPype 0.8. 
>Unfortunately, I have run into an issue that would break compatible. 
>Currently boxed types inherit from primitives so that they will
>function anywhere that requires a number type without requiring a
>forced conversion.  This is forcing everything is to go through a type
>system which pulls in a huge number of python calls for each routine. 
>I can remove this restriction, but the isinstance will no longer work
>for boxed types.  I will have to make them work like numbers rather
>than being int or floats.
>
>Any opinions?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/jpype-project/jpype/issues/536

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by Thrameos** [comment]
Okay i will give it a try.

**Comment by Thrameos** [comment]
@marscher I tried the API and got it to work but I will need some additional changes to use it.  I will need to have at least two base classes because in addition to the boxed types, exceptions also need a different base.  Most classes can be derived from PyJPValue, but others will need to be heap type derived from long, double, and exception.  The docs on heap types are a bit thin.

But this raises something I hadn't considered before.  If we can fake the inheritance tree why do we need to create the whole Java tree?  The single biggest bootstrapping problem is to create one class I have to build wrappers for every super class and every interface that are in the inheritance tree.  The downside is that rather than using inheritance to get the fields and methods, I would have to create all the fields and methods at each wrapper class.  It would be a bit more memory.  On the other hand it would be faster as every method lookup would be O(1).  

The other problem that I encountered is that if I use this API, that means adding a Python call for every isinstance call.  I could improve it slightly by implementing it in C. 
Unfortunately, we do a huge number of these because every parameter lookup requires one of these for each argument for each overload.  (This is currently a problem because our tree is very deep which is why I went with the ``__javavalue__`` solution.)  Thus to make use of this API we need a faster method of getting the base type.  Looking through cpython source code I see one trick that may make it possible.  The Python type system copies the values of any functions that were not overridden (assuming it is an inherited method).  In a few places rather than using the slot directly it checks to see if the slot matches a predefined function and instead calls a different function.  We could use this same check to convert all the determinations of java values into O(1).  If I check if the ``tp_getattr`` points to ``PyJPValue::get`` then I would immediately know that it contains ``PyJPValue`` is the base class.  It seems a bit hackish, but it would speed up the process a bit. 

Does this sound like a good tradeoff?
